### PR TITLE
Treasury synthesis

### PR DESCRIPTION
**Pull request by DrPresident**
This is a rough implementation that needs further testing. It is not feature complete and will need further iterations.

### CODE REVIEW

File: contracts/treasury/src/handle.rs [hunk 1]
```diff
@@ -1,43 +1,119 @@
...
+
+    allocations_w(&mut deps.storage).update(
+        asset.contract.address.to_string().as_bytes(),
+        |allocs| {
+            let mut alloc_list = match allocs {
+                None => {
+                    vec![]
+                }
+                Some(a) => a,
+            };
```
Comment (FloppyDisck) [inline contracts/treasury/src/handle.rs:55]: You can clean up all of the matches like this

```suggestion
            let mut alloc_list = allocs.unwrap_or(vec![]);
```

File: contracts/treasury/src/handle.rs [hunk 2]
```diff
@@ -125,15 +207,209 @@ pub fn try_register_asset<S: Storage, A: Api, Q: Querier>(
...
+    env: &Env,
+    asset: HumanAddr,
+    alloc: Allocation,
 ) -> StdResult<HandleResponse> {
-    let messages = vec![];
+    let config = config_r(&deps.storage).load()?;
+
+    /* ADMIN ONLY */
+    if env.message.sender != config.admin {
+        return Err(StdError::Unauthorized { backtrace: None });
```
Comment (FloppyDisck) [inline contracts/treasury/src/handle.rs:220]:
```suggestion
        return Err(StdError::unauthorized());
```

File: contracts/treasury/src/handle.rs [hunk 3]
```diff
@@ -125,15 +207,209 @@ pub fn try_register_asset<S: Storage, A: Api, Q: Querier>(
...
+    }
+
+    let full_asset = match assets_r(&deps.storage).may_load(asset.to_string().as_bytes())? {
+        Some(a) => a,
+        None => {
+            return Err(StdError::GenericErr {
+                msg: "Unregistered asset".to_string(),
+                backtrace: None,
+            });
```
Comment (FloppyDisck) [inline contracts/treasury/src/handle.rs:229]: You can make these errors cleaner like this

```suggestion
            return Err(StdError::generic_err ("Unregistered asset"));
```

File: packages/shade_protocol/src/task.rs [hunk 4]
```diff
@@ -0,0 +1,92 @@
+use cosmwasm_std::{HumanAddr, Decimal, Uint128};
```
Comment (FloppyDisck) [inline packages/shade_protocol/src/task.rs:1]: You can probably generalize this using the admin command style that governance uses. Ill probably work on this in the week to get all of the airdrop stuff out of the way

File: packages/network_tester/src/scrt_staking.rs [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+        if let micro_mint::QueryAnswer::MintLimit { limit } = query {
+            println!("Mint limit after change request");
+            print_struct(limit);
+        }
+    }
+    */
+
+    Ok(())
+}
+
```
Comment (FloppyDisck) [inline packages/network_tester/src/scrt_staking.rs:346]: look into ```tests/testnet_integration.rs```. These are now stored in their own functions

File: packages/network_tester/src/scrt_staking.rs [hunk 6]
```diff
@@ -0,0 +1,412 @@
+use colored::*;
```
Comment (FloppyDisck) [inline packages/network_tester/src/scrt_staking.rs:1]: try moving this into tests and making it a test function so it gets automatically run by the tester

**Review by FloppyDisck — APPROVED** [review]
Would be nice if you could look into the comments, but everything looks good aside from that!

File: packages/shade_protocol/src/task.rs (hunk 4 above)
Comment (DrPresident) [reply packages/shade_protocol/src/task.rs:1]: ugh none of the `task` stuff was supposed to be in there, I was just fiddling around seeing what would work. I'm gonna go cut out all that stuff.

**Pull request merged**